File: Stack/generate-parentheses/generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
    const stack = [];
    const res = [];

    const backtrack = (openN, closeN) => {
        if (openN === closeN && openN === n) {
            res.push(stack.join(""));
            return;
        }

        if (openN < n) {
            stack.push("(");
            backtrack(openN + 1, closeN);
            stack.pop();
        }

        if (closeN < openN) {
            stack.push(")");
            backtrack(openN, closeN + 1);
            stack.pop();
        }
    };

    backtrack(0, 0);
    return res;
};

File: Stack/generate-parentheses/generate-parentheses.cs
public class Solution
{
    public IList<string> GenerateParenthesis(int n)
    {
        List<char> stack = new List<char>();
        List<string> res = new List<string>();

        void Backtrack(int openN, int closeN)
        {
            if (openN == closeN && openN == n)
            {
                res.Add(new string(stack.ToArray()));
                return;
            }

            if (openN < n)
            {
                stack.Add('(');
                Backtrack(openN + 1, closeN);
                stack.RemoveAt(stack.Count - 1);
            }

            if (closeN < openN)
            {
                stack.Add(')');
                Backtrack(openN, closeN + 1);
                stack.RemoveAt(stack.Count - 1);
            }
        }

        Backtrack(0, 0);
        return res;
    }
}